functions/go/gatekeeper-validate/main.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"os"
+
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/gatekeeper-validate/generated"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+func main() {
+	resourceList := &framework.ResourceList{}
+
+	cmd := framework.Command(resourceList, func() error {
+		var objects []runtime.Object
+		for _, item := range resourceList.Items {
+			meta, err := item.GetValidatedMetadata()
+			if err != nil {
+				return err
+			}
+
+			s, err := item.String()
+			if err != nil {
+				return err
+			}
+			obj, err := scheme.New(schema.FromAPIVersionAndKind(meta.APIVersion, meta.Kind))
+			switch {
+			case runtime.IsNotRegisteredError(err):
+				obj = &unstructured.Unstructured{}
+			case err != nil:
+				return err
+			}
+			err = k8syaml.Unmarshal([]byte(s), obj)
+			if err != nil {
+				return err
+			}
+			objects = append(objects, obj)
+		}
+
+		err := Validate(objects)
+		if err == nil {
+			return nil
+		}
+
+		if result, ok := err.(*framework.Result); ok {
```

Comment:
What happens if the typecast fails, which seems likely to be the case if this code from parseResults is hit?

```
		if !ok {
			return fmt.Errorf("could not cast to unstructured: %+v", r.Resource)
		}
```

Instead of returning an error which can then potentially be cast, maybe always return a list of results?

The filtering function seems like a good idea. I'm not sure what the significance of the return value is here (does it fail differently?)

Reply:
Looking at the code, it does seem functionally correct, but avoiding the typecast seems both simpler and less brittle.